File: tests/test_sym_representation.cpp
/*
 * Copyright (c) 2022 Authors:
 *   - Trevor Phillips <trevphil3@gmail.com>
 *
 * Distributed under the CC BY-NC-SA 4.0 license
 * (See accompanying file LICENSE.md).
 */

#include <gtest/gtest.h>

#include <filesystem>
#include <vector>

#include "core/logic_gate.hpp"
#include "core/sym_representation.hpp"

namespace preimage {

TEST(SymRepresentationTest, Initialization) {
  const std::vector<LogicGate> gates = {LogicGate(LogicGate::Type::and_gate, 4, {1, -2}),
                                        LogicGate(LogicGate::Type::xor_gate, 5, {2, -3})};
  SymRepresentation rep(gates, {1, 2, 3}, {0, 4, 5});
  EXPECT_EQ(rep.numVars(), 5);
  EXPECT_EQ(rep.gates().size(), 2);

  const std::vector<int> inputs = {1, 2, 3};
  const std::vector<int> outputs = {0, 4, 5};
  EXPECT_EQ(rep.inputIndices(), inputs);
  EXPECT_EQ(rep.outputIndices(), outputs);
}

TEST(SymRepresentationTest, PruneAndReindex) {
  const std::vector<LogicGate> gates = {LogicGate(LogicGate::Type::and_gate, 4, {1, -2}),
                                        LogicGate(LogicGate::Type::and_gate, 5, {3, -4})};
  SymRepresentation rep(gates, {1, 2, 3}, {4});
  EXPECT_EQ(rep.numVars(), 3);
  EXPECT_EQ(rep.gates().size(), 1);
  EXPECT_EQ(rep.gates().at(0).t(), LogicGate::Type::and_gate);

  const std::vector<int> inputs = {1, 2, 0};
  const std::vector<int> outputs = {3};
  EXPECT_EQ(rep.inputIndices(), inputs);
  EXPECT_EQ(rep.outputIndices(), outputs);
}

TEST(SymRepresentationTest, ConvertToCNF) {
  const std::vector<LogicGate> gates = {LogicGate(LogicGate::Type::and_gate, 4, {1, -2}),
                                        LogicGate(LogicGate::Type::and_gate, 5, {3, -4})};
  SymRepresentation rep(gates, {1, 2, 3}, {4});
  const CNF cnf = rep.toCNF();
  EXPECT_EQ(cnf.num_vars, 3);
  EXPECT_EQ(cnf.num_clauses, 3);
}

TEST(SymRepresentationTest, ConvertDAG) {
  const std::vector<LogicGate> gates = {
      LogicGate(LogicGate::Type::and_gate, 4, {1, -2}),
      LogicGate(LogicGate::Type::xor_gate, 5, {2, -3}),
      LogicGate(LogicGate::Type::maj3_gate, 6, {1, 4, 5})};
  const std::vector<int> inputs = {1, 2, 3};
  const std::vector<int> outputs = {0, 6, 5, 0, 0};
  SymRepresentation rep(gates, inputs, outputs);
  EXPECT_EQ(rep.numVars(), 6);
  EXPECT_EQ(rep.gates().size(), 3);
  EXPECT_EQ(rep.inputIndices(), inputs);
  EXPECT_EQ(rep.outputIndices(), outputs);

  const std::filesystem::path path = std::filesystem::temp_directory_path() / "dag.txt";
  rep.toDAG(path.string());
  rep = SymRepresentation::fromDAG(path.string());
  EXPECT_EQ(rep.numVars(), 6);
  EXPECT_EQ(rep.gates().size(), 3);
  EXPECT_EQ(rep.inputIndices(), inputs);
  EXPECT_EQ(rep.outputIndices(), outputs);
}

TEST(SymRepresentationTest, LoadInvalidDAG) {
  std::filesystem::path path = std::filesystem::temp_directory_path();
  path = path / "not" / "a" / "dag.txt";
  EXPECT_ANY_THROW({ SymRepresentation::fromDAG(path.string()); });
}

}  // end namespace preimage

File: cryptosym/tests/test_sym_representation.py
"""
Copyright (c) 2022 Authors:
    - Trevor Phillips <trevphil3@gmail.com>

Distributed under the CC BY-NC-SA 4.0 license
(See accompanying file LICENSE.md).
"""

from pathlib import Path

import pytest

from cryptosym import LogicGate, SymRepresentation

GType = LogicGate.Type


class TestSymRepresentation:
    def test_initialization(self):
        gates = [
            LogicGate(GType.and_gate, 4, [1, -2]),
            LogicGate(GType.xor_gate, 5, [2, -3]),
        ]
        rep = SymRepresentation(
            gates=gates, input_indices=[1, 2, 3], output_indices=[0, 4, 5]
        )
        assert rep.num_vars == 5
        assert len(rep.gates) == 2
        assert rep.input_indices == [1, 2, 3]
        assert rep.output_indices == [0, 4, 5]

    def test_set_readonly_properties(self):
        gates = [
            LogicGate(GType.and_gate, 4, [1, -2]),
            LogicGate(GType.xor_gate, 5, [2, -3]),
        ]
        rep = SymRepresentation(gates, [1, 2, 3], [0, 4, 5])
        with pytest.raises(AttributeError):
            rep.num_vars = 10
        with pytest.raises(AttributeError):
            rep.input_indices = [10, 20, 30]
        with pytest.raises(AttributeError):
            rep.output_indices = [10, 20, 30]
        with pytest.raises(AttributeError):
            rep.gates = []

    def test_prune_and_reindex(self):
        gates = [
            LogicGate(GType.and_gate, 4, [1, -2]),
            LogicGate(GType.and_gate, 5, [3, -4]),
        ]
        rep = SymRepresentation(gates, [1, 2, 3], [4])
        assert rep.num_vars == 3
        assert len(rep.gates) == 1
        assert rep.gates[0].t == GType.and_gate
        assert rep.input_indices == [1, 2, 0]
        assert rep.output_indices == [3]

    def test_convert_to_cnf(self):
        gates = [
            LogicGate(GType.and_gate, 4, [1, -2]),
            LogicGate(GType.and_gate, 5, [3, -4]),
        ]
        rep = SymRepresentation(gates, [1, 2, 3], [4])
        cnf = rep.to_cnf()
        assert cnf.num_vars == 3
        assert cnf.num_clauses == 3

    def test_convert_dag(self, tmp_path: Path):
        gates = [
            LogicGate(GType.and_gate, 4, [1, -2]),
            LogicGate(GType.xor_gate, 5, [2, -3]),
            LogicGate(GType.maj3_gate, 6, [1, 4, 5]),
        ]
        inputs = [1, 2, 3]
        outputs = [0, 6, 5, 0, 0]
        rep = SymRepresentation(gates, inputs, outputs)
        assert rep.num_vars == 6
        assert len(rep.gates) == 3
        assert rep.input_indices == inputs
        assert rep.output_indices == outputs

        rep.to_dag(tmp_path / "dag.txt")
        loaded = SymRepresentation(dag_file=tmp_path / "dag.txt")
        assert loaded.num_vars == 6
        assert len(loaded.gates) == 3
        assert loaded.input_indices == inputs
        assert loaded.output_indices == outputs

    def test_load_invalid_dag(self):
        with pytest.raises(ValueError):
            _ = SymRepresentation("does_not_exist.txt")
        with pytest.raises(ValueError):
            _ = SymRepresentation(Path("does_not_exist.txt"))
